videostore/indexer.go

```diff
@@ -0,0 +1,406 @@
+	return nil
+}
+
+// RemoveIndexedFiles removes file paths from the index.
+// This function acquires the indexer's mutex.
+func (ix *Indexer) RemoveIndexedFiles(filePaths []string) error {
+	ix.mu.Lock()
+	defer ix.mu.Unlock()
+
+	if len(filePaths) == 0 {
+		return nil
+	}
+
+	tx, err := ix.db.Begin()
+	if err != nil {
+		return fmt.Errorf("failed to begin transaction: %w", err)
+	}
+	defer tx.Rollback() // rollback on failure
+
+	query := fmt.Sprintf("DELETE FROM %s WHERE file_path = ?", segmentsTableName)
+	stmt, err := tx.Prepare(query)
+	if err != nil {
+		return fmt.Errorf("failed to prepare delete statement: %w", err)
+	}
+	defer stmt.Close()
+
+	for _, path := range filePaths {
+		if _, err := stmt.Exec(path); err != nil {
+			return fmt.Errorf("failed to execute delete for path '%s': %w", path, err)
+		}
+		ix.logger.Debugf("removed indexed segment: %s", path)
+	}
+
+	return tx.Commit()
+}
+
+// getVideoDuration uses the getVideoInfo utility to get the duration of a video file.
+// This function does not require the indexer mutex.
+func (ix *Indexer) getVideoDuration(filePath string) (int64, error) {
+	info, err := getVideoInfo(filePath)
+	if err != nil {
+		return 0, fmt.Errorf("getVideoInfo failed for %s: %w", filePath, err)
+	}
+	// getVideoInfo returns duration in microseconds, convert to milliseconds
+	durationMs := info.duration.Milliseconds()
+	return durationMs, nil
+}
+
+// timeRange represents a single contiguous block of stored video.
+type timeRange struct {
+	Start time.Time
+	End   time.Time
+}
+
+// StorageState summarizes the state of the stored video segments.
+type StorageState struct {
+	TotalSizeBytes  int64
+	TotalDurationMs int64
+	SegmentCount    int
+	Ranges          []timeRange // List of contiguous time ranges
+}
+
+// GetStorageState calculates the current storage state from the index.
+// This function acquires the indexer's mutex.
+func (ix *Indexer) GetStorageState() (StorageState, error) {
+	ix.mu.Lock()
+	defer ix.mu.Unlock()
+
+	var state StorageState
+
+	// Query all segments, ordered by start time
+	query := fmt.Sprintf(`
+	SELECT start_time_unix, duration_ms, size_bytes
+	FROM %s
+	ORDER BY start_time_unix ASC
+	`, segmentsTableName)
+
+	rows, err := ix.db.Query(query)
+	if err != nil {
+		return StorageState{}, fmt.Errorf("failed to query segments for state: %w", err)
+	}
+	defer rows.Close()
+
+	var ranges []timeRange
+	var currentRange *timeRange
+	const slopDuration = 5 * time.Second
+
+	for rows.Next() {
+		var startTimeUnix int64
+		var durationMs sql.NullInt64
+		var sizeBytes sql.NullInt64
+
+		if err := rows.Scan(&startTimeUnix, &durationMs, &sizeBytes); err != nil {
+			return StorageState{}, fmt.Errorf("failed to scan segment row: %w", err)
+		}
+
+		state.SegmentCount++
+		if durationMs.Valid {
+			state.TotalDurationMs += durationMs.Int64
+		} else {
+			ix.logger.Warnw("segment found with NULL duration in index", "start_time_unix", startTimeUnix)
+			continue
```

Comment:
have you observed this condition? When does it happen?

Reply:
Same thing as https://github.com/viam-modules/video-store/pull/65#discussion_r2086864562

>I haven't seen it happen. I thought it would be best practice to case it out.